core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala

```diff
@@ -0,0 +1,327 @@
+  def testZkAntiMigration() {
+    val unsecureZkUtils = ZkUtils(zkConnect, 6000, 6000, false)
+    try {
+      testMigration(zkUtils, unsecureZkUtils)
+    } finally {
+      unsecureZkUtils.close()
+    }
+  }
+
+  /**
+   * Tests that the persistent paths cannot be deleted.
+   */
+  @Test
+  def testDelete() {
+    info("zkConnect string: %s".format(zkConnect))
+    ZkSecurityMigrator.run(Array("--zookeeper.acl=secure", "--zookeeper.connect=%s".format(zkConnect)))
+    deleteAllUnsecure()
+  }
+
+  /**
+   * Tests that znodes cannot be deleted when the 
+   * persistent paths have children.
+   */
+  @Test
+  def testDeleteRecursive() {
+    info("zkConnect string: %s".format(zkConnect))
+    for (path <- zkUtils.securePersistentZkPaths) {
+      info("Creating " + path)
+      zkUtils.makeSurePersistentPathExists(path)
+      zkUtils.createPersistentPath(path + "/fpjwashere", "")
+    }
+    zkUtils.zkConnection.setAcl("/", zkUtils.DefaultAcls, -1)
+    deleteAllUnsecure()
+  }
+  
+  /**
+   * Tests the migration tool when chroot is being used.
+   */
+  @Test
+  def testChroot {
+    zkUtils.createPersistentPath("/kafka")
+    val unsecureZkUtils = ZkUtils(zkConnect + "/kafka", 6000, 6000, false)
+    val secureZkUtils = ZkUtils(zkConnect + "/kafka", 6000, 6000, true)
+    try {
+      testMigration(unsecureZkUtils, secureZkUtils)
+    } finally {
+      unsecureZkUtils.close()
+      secureZkUtils.close()
+    }
+  }
+
+  /**
+   * Exercises the migration tool. It is used by two test cases:
+   * testZkMigration and testZkAntiMigration.
+   */
+  private def testMigration(firstZk: ZkUtils, secondZk: ZkUtils) {
+    info("zkConnect string: %s".format(zkConnect))
+    for (path <- firstZk.securePersistentZkPaths) {
+      info("Creating " + path)
+      firstZk.makeSurePersistentPathExists(path)
+      // Create a child for each znode to exercise the recurrent
+      // traversal of the data tree
+      firstZk.createPersistentPath(path + "/fpjwashere", "")
+    }
+    val secureOpt: String  = secondZk.isSecure match {
+      case true => "secure"
+      case false => "unsecure"
+    }
+    ZkSecurityMigrator.run(Array("--zookeeper.acl=%s".format(secureOpt), "--zookeeper.connect=%s".format(zkConnect)))
+    info("Done with migration")
+    for (path <- secondZk.securePersistentZkPaths) {
+      val listParent = (secondZk.zkConnection.getAcl(path)).getKey
+      assertTrue(path, isAclCorrect(listParent, secondZk.isSecure))
+
+      val childPath = path + "/fpjwashere"
+      val listChild = (secondZk.zkConnection.getAcl(childPath)).getKey
+      assertTrue(childPath, isAclCorrect(listChild, secondZk.isSecure))
+    }
+  }
+
+  /**
+   * Verifies that the path has the appropriate secure ACL.
+   */
+  private def verify(path: String): Boolean = {
+    val list = (zkUtils.zkConnection.getAcl(path)).getKey
+    var result: Boolean = true
+    for (acl: ACL <- list.asScala) {
+           result = result && isAclSecure(acl)
+    }
+    result
```

Comment:
@ijuma Maybe. The logic needs to be such that the outcome is false if at least one of the calls to `isAclSecure(acl)`.